### DISCUSSION TITLE
Map DBAPIError back to a specific engine

### DISCUSSION THREAD

**Discussion by tonycosentini** [sqlalchemy/sqlalchemy#11744] [Usage Questions] [answered]
Is there any way to map a `DBAPIError` back to the specific engine/connection that caused it?

I'm trying to implement some basic health traffic for different engine/session pairs in our app, but it seems like it's not really possible to map errors back to an underlying connection.

Thanks!

**Comment by zzzeek** [comment, marked as answer]
the engine itself isn't passed along to the SQLAlchemy exception objects but typically tracing out the cause of an exception is via inspection of the stack trace.  it would show you the exact line of code at which the exception originated and by looking at the objects in use there you'd know what engine you're dealing with; unless you have some kind of dynamic on-the-fly engine thing going on.

if you have some kind of dynamic engine thing, like you're sharding or something like that, then you probably want to use [Exception.add_note](https://peps.python.org/pep-0678/) in combination with SQLAlchemy's [handle_error](https://docs.sqlalchemy.org/en/20/core/events.html#sqlalchemy.events.DialectEvents.handle_error) event hook.   [ExceptionContext](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.ExceptionContext) gives you everything to know about the error being raised and you can augment the SQLAlchemy exception right there or rewrite the message, whatever you want.

**Reply by tonycosentini** [reply]
This is perfect, thanks as always @zzzeek
